pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go

```diff
@@ -0,0 +1,267 @@
+
+	log = r.log.With("cluster", cluster.Name, "restore", restore.Name)
+
+	if cluster.Labels[kubermaticv1.WorkerNameLabelKey] != r.workerName {
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, log, restore, cluster)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(restore, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+		r.recorder.Eventf(cluster, corev1.EventTypeWarning, "ReconcilingError",
+			"failed to reconcile etcd restore %q: %v", restore.Name, err)
+	}
+
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, restore *kubermaticv1.EtcdRestore, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	log.Infof("performing etcd restore from backup %v", restore.Spec.BackupName)
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseCompleted {
+		return nil, nil
+	}
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseStsRebuilding {
+		return r.rebuildEtcdStatefulset(ctx, log, restore, cluster)
+	}
+
+	// pause cluster
+	if err := r.updateCluster(ctx, cluster, func(cluster *kubermaticv1.Cluster) {
+		cluster.Spec.Pause = true
+	}); err != nil {
+		return nil, fmt.Errorf("failed to pause cluster: %v", err)
+	}
+
+	if err := r.updateRestore(ctx, restore, func(restore *kubermaticv1.EtcdRestore) {
+		restore.Status.Phase = kubermaticv1.EtcdRestorePhaseStarted
+		kuberneteshelper.AddFinalizer(restore, RebuildStatefulsetFinalizer)
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add finalizer: %v", err)
+	}
+
+	// delete etcd sts
+	sts := &v1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: cluster.Status.NamespaceName, Name: resources.EtcdStatefulSetName}, sts)
+	if err == nil {
+		if err := r.Delete(ctx, sts); err != nil && !kerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to delete etcd statefulset: %v", err)
+		}
+	} else if !kerrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get etcd statefulset: %v", err)
+	}
+
+	// delete PVCs
+	pvcSelector, err := labels.Parse(fmt.Sprintf("%s=%s", resources.AppLabelKey, resources.EtcdStatefulSetName))
+	if err != nil {
+		return nil, fmt.Errorf("software bug: failed to parse etcd pvc selector: %v", err)
+	}
+
+	pvcs := &corev1.PersistentVolumeClaimList{}
+	if err := r.List(ctx, pvcs, &ctrlruntimeclient.ListOptions{Namespace: cluster.Status.NamespaceName, LabelSelector: pvcSelector}); err != nil {
+		return nil, fmt.Errorf("failed to list pvcs (%v): %v", pvcSelector.String(), err)
+	}
+
+	for _, pvc := range pvcs.Items {
+		deletePropagationForeground := metav1.DeletePropagationForeground
+		delOpts := &ctrlruntimeclient.DeleteOptions{
+			PropagationPolicy: &deletePropagationForeground,
+		}
+		if err := r.Delete(ctx, &pvc, delOpts); err != nil {
+			return nil, fmt.Errorf("failed to delete pvc %v: %v", pvc.GetName(), err)
+		}
+	}
+
+	if len(pvcs.Items) > 0 {
```

Comment:
I think we should List() again before this?

Reply:
Since we just deleted at least one PVC (as evidenced by `len(pvcs.Items) > 0`) and the deletion happens asynchronously in the background, I wouldn't expect the PVC to be gone immediately afterwards (and wouldn't it still be in the cache anyway?). That's why I'm just arranging for the reconcile to be called again 30 seconds later, and this repeats until no PVCs are found anymore.